draft-ietf-mls-protocol.md

```diff
@@ -2883,93 +2882,97 @@ def level(x):
         k += 1
     return k
 
-# The number of nodes needed to represent a tree with n leaves
+# The number of nodes needed to represent a tree with n leaves.
 def node_width(n):
     return 2*(n - 1) + 1
 
-# The index of the root node of a tree with n leaves
+# The index of the root node of a tree with n leaves.
 def root(n):
     w = node_width(n)
     return (1 << log2(w)) - 1
 
-# The left child of an intermediate node.  Note that because the
-# tree is left-balanced, there is no dependency on the size of the
-# tree.  The child of a leaf node is itself.
+# The left child of an intermediate node. Note that because the tree is
+# left-balanced, there is no dependency on the size of the tree. The
+# child of a leaf node is itself.
 def left(x):
     k = level(x)
     if k == 0:
-        return x
+        raise Exception('leaf node has no children')
 
     return x ^ (0x01 << (k - 1))
 
-# The right child of an intermediate node.  Depends on the size of
-# the tree because the straightforward calculation can take you
-# beyond the edge of the tree.  The child of a leaf node is itself.
+# The right child of an intermediate node. Depends on the number of
+# leaves because the straightforward calculation can take you beyond the
+# edge of the tree.
 def right(x, n):
     k = level(x)
     if k == 0:
-        return x
+        raise Exception('leaf node has no children')
 
     r = x ^ (0x03 << (k - 1))
     while r >= node_width(n):
         r = left(r)
     return r
 
-# The immediate parent of a node.  May be beyond the right edge of
-# the tree.
+# The immediate parent of a node. May be beyond the right edge of the
+# tree.
 def parent_step(x):
     k = level(x)
     b = (x >> (k + 1)) & 0x01
     return (x | (1 << k)) ^ (b << (k + 1))
 
-# The parent of a node.  As with the right child calculation, have
-# to walk back until the parent is within the range of the tree.
+# The parent of a node. As with the right child calculation, we have to
+# walk back until the parent is within the range of the tree.
 def parent(x, n):
     if x == root(n):
-        return x
+        raise Exception('root node has no parent')
 
     p = parent_step(x)
     while p >= node_width(n):
         p = parent_step(p)
     return p
 
-# The other child of the node's parent.  Root's sibling is itself.
+# The other child of the node's parent.
 def sibling(x, n):
     p = parent(x, n)
     if x < p:
         return right(p, n)
-    elif x > p:
+    else:
         return left(p)
 
-    return p
-
-# The direct path of a node, ordered from the root
-# down, not including the root or the terminal node
+# The direct path of a node, ordered from leaf to root.
```

Comment:
```suggestion
# The direct path of a node, ordered from leaf to root,
# including the root but not the leaf.
```

Reply:
This is redundant with the definition of direct path